subworkflow patho_workflow: 
    workdir: "pathogenicityScores"
    snakefile: "pathogenicityScores/Snakefile"

subworkflow population_workflow: 
    workdir: "populationDBs"
    snakefile: "populationDBs/Snakefile"

subworkflow reference_workflow: 
    workdir: "referenceGenome"
    snakefile: "referenceGenome/Snakefile"

subworkflow jannovar_workflow: 
    workdir: "jannovar"
    snakefile: "jannovar/Snakefile"

subworkflow omim_workflow: 
    workdir: "omim"
    snakefile: "omim/Snakefile"


# download the data!
rule all:
    input:
        patho_workflow("cadd_exon_snv.v1.4.tsv.gz"),
        patho_workflow("cadd_exon_snv.v1.4.tsv.gz.tbi"),
        patho_workflow("cadd_exon_indel.v1.4.tsv.gz"),
        patho_workflow("cadd_exon_indel.v1.4.tsv.gz.tbi"),
        population_workflow("1KG_ncbi_exon.vcf.gz"),
        population_workflow("1KG_ncbi_exon.vcf.gz.tbi"),
        population_workflow("ExAC.r1.sites.vep.vcf.gz"),
        population_workflow("ExAC.r1.sites.vep.vcf.gz.tbi"),
        population_workflow("UK10K_ncbi_exon.vcf.gz"),
        population_workflow("UK10K_ncbi_exon.vcf.gz.tbi"),
        reference_workflow("data/human_g1k_v37.fasta"),
        reference_workflow("data/human_g1k_v37.fasta.fai"),
        reference_workflow("data/human_g1k_v37.fasta.dict"),
        jannovar_workflow("data/hg19_refseq.ser"),
        jannovar_workflow("jannovar_0.25/data/hg19_refseq.ser"),
        jannovar_workflow("jannovar_0.26/data/hg19_refseq.ser"),
        omim_workflow("genemap2.txt"),
        omim_workflow("mim2gene.txt"),
        omim_workflow("mim_to_ps.json"),
        omim_workflow("mimTitles.txt"),
        omim_workflow("morbidmap.txt"),
        omim_workflow("omim_deprecated_replacement.json"),
        omim_workflow("phenotypicSeries.txt"),

